main/streams/userspace.c

```diff
@@ -836,205 +812,246 @@ static void statbuf_from_array(zval *array, php_stream_statbuf *ssb)
 
 static int php_userstreamop_stat(php_stream *stream, php_stream_statbuf *ssb)
 {
-	zval func_name;
 	zval retval;
-	int call_result;
 	php_userstream_data_t *us = (php_userstream_data_t *)stream->abstract;
 	int ret = -1;
 
-	ZVAL_STRINGL(&func_name, USERSTREAM_STAT, sizeof(USERSTREAM_STAT)-1);
+	zend_string *func_name = ZSTR_INIT_LITERAL(USERSTREAM_STAT, false);
+	zend_result call_result = zend_call_method_if_exists(Z_OBJ(us->object), func_name, &retval, 0, NULL);
+	zend_string_release_ex(func_name, false);
 
-	call_result = call_method_if_exists(&us->object, &func_name, &retval, 0, NULL);
+	if (UNEXPECTED(call_result == FAILURE)) {
+		php_error_docref(NULL, E_WARNING, "%s::" USERSTREAM_STAT " is not implemented!",
+				ZSTR_VAL(us->wrapper->ce->name));
+		return -1;
+	}
+	if (UNEXPECTED(Z_ISUNDEF(retval))) {
+		return -1;
+	}
 
-	if (call_result == SUCCESS && Z_TYPE(retval) == IS_ARRAY) {
+	if (EXPECTED(Z_TYPE(retval) == IS_ARRAY)) {
 		statbuf_from_array(&retval, ssb);
 		ret = 0;
-	} else {
-		if (call_result == FAILURE) {
-			php_error_docref(NULL, E_WARNING, "%s::" USERSTREAM_STAT " is not implemented!",
-					ZSTR_VAL(us->wrapper->ce->name));
-		}
 	}
+	// TODO: Warning on incorrect return type?
 
 	zval_ptr_dtor(&retval);
-	zval_ptr_dtor(&func_name);
 
 	return ret;
 }
 
-
-static int php_userstreamop_set_option(php_stream *stream, int option, int value, void *ptrparam) {
-	zval func_name;
+static int user_stream_set_check_liveliness(const php_userstream_data_t *us)
+{
 	zval retval;
-	int call_result;
-	php_userstream_data_t *us = (php_userstream_data_t *)stream->abstract;
-	int ret = PHP_STREAM_OPTION_RETURN_NOTIMPL;
-	zval args[3];
 
-	switch (option) {
-	case PHP_STREAM_OPTION_CHECK_LIVENESS:
-		ZVAL_STRINGL(&func_name, USERSTREAM_EOF, sizeof(USERSTREAM_EOF)-1);
-		call_result = call_method_if_exists(&us->object, &func_name, &retval, 0, NULL);
-		if (call_result == SUCCESS && (Z_TYPE(retval) == IS_FALSE || Z_TYPE(retval) == IS_TRUE)) {
-			ret = zval_is_true(&retval) ? PHP_STREAM_OPTION_RETURN_ERR : PHP_STREAM_OPTION_RETURN_OK;
-		} else {
-			ret = PHP_STREAM_OPTION_RETURN_ERR;
-			php_error_docref(NULL, E_WARNING,
-					"%s::" USERSTREAM_EOF " is not implemented! Assuming EOF",
-					ZSTR_VAL(us->wrapper->ce->name));
-		}
+	zend_string *func_name = ZSTR_INIT_LITERAL(USERSTREAM_EOF, false);
+	zend_result call_result = zend_call_method_if_exists(Z_OBJ(us->object), func_name, &retval, 0, NULL);
+	zend_string_release_ex(func_name, false);
+
+	if (UNEXPECTED(call_result == FAILURE)) {
+		php_error_docref(NULL, E_WARNING,
+				"%s::" USERSTREAM_EOF " is not implemented! Assuming EOF",
+				ZSTR_VAL(us->wrapper->ce->name));
+		// TODO: Returning PHP_STREAM_OPTION_RETURN_NOTIMPL means the stream_set_option() returns true...
```

Comment:
I don't think it should return `PHP_STREAM_OPTION_RETURN_NOTIMPL` in this case as it has got slightly different meaning for options. It's more for streams where such options has not an important meaning but in this case liveness should be implemented because it might be important for user stream.